### PR TITLE
Sanitize user params and clean up input

### PR DESCRIPTION
**Pull request by apavlen**
# Pull Request

## Title

_Add validation and default settings to ClusterStateConfig for required parameters_

---

## Description

_From the discussion in https://github.com/microsoft/vasim/pull/29#discussion_r1742436854 we need to sanitize user params (ex: window, lag) so that things don't break unexpectedly. We add some basic validations and default settings. _

- **Issue link**: (optional) _#31_

---

## Type of Change

_Indicate the type of change by choosing one (or more) of the following:_

- 🔄 Refactor  
- 🧪 Tests

---

## Testing

_Unit tests and notebook_

---

## Additional Notes (optional)

_As part of this change also added field "enabled" for future use is https://github.com/microsoft/vasim/issues/42.
if enabled is False, prediction config is validated and defaults are added. Otherwise it's ignored. We will add more checks for validation and change code for prediction separately_

### CODE REVIEW

**Review by ksaur — CHANGES_REQUESTED** [review]
One thing - in the notebook there is a section _"prediction_config : this relates to the window of...."_  where it describes the prediction config and each field. Can you please add a line about what the new `"enabled": true,`does? Thanks!

**Comment by ksaur** [comment]
I think this also closes  https://github.com/microsoft/vasim/issues/70 ?

File: examples/using_vasim.ipynb [hunk 1]
```diff
@@ -171,13 +171,14 @@
...
     "  * `lag` (float): Number of minutes to wait after making a decision\n",
     "  * `min_cpu_limit` (int): The minimum number of cores to recommend. This is used as a [safety guard](https://github.com/microsoft/vasim/blob/198a06062a91f6455b87710b0e59834530b6ea29/simulator/SimulatedInfraScaler.py#L56) in the simulated infra.\n",
     "  * `max_cpu_limit` (int): The maximum number of cores to recommend. (same as min)\n",
     "  * `recovery_time` (int of minutes, optional, default is 15): The **minimum** amount of time in minutes to wait before making another scaling decision.\n",
     "* **`prediction_config`** : this relates to the window of [predicted](https://github.com/microsoft/vasim/blob/198a06062a91f6455b87710b0e59834530b6ea29/recommender/cluster_state_provider/PredictiveFileClusterStateProvider.py#L29) data that is fed into the algorithm. It uses a time series to forecast what the data might be in the future to help the algorithm proactively scale\n",
+    "  * `enabled`: (boolean): This is a flag that enables prediction configuration. If it's set to false, prediction is disabled.  \n",
```
Comment (apavlen) [inline examples/using_vasim.ipynb:181]: @ksaur added

File: examples/using_vasim.ipynb [hunk 2]
```diff
@@ -171,13 +171,14 @@
     "There are three sets of parameters:\n",
     "* **`algo_specific_config`** : this is where you put the parameters you will use in your algorithm's `run` function\n",
     "  * An example: `addend: 2` for the `DummyAdditiveRecommender` we'll be using for this example.\n",
-    "* **`general_config`** : These are the parameters related to how the csv trace data is passed in, and the simulation safe-guards.\n",
+    "* **`general_config`** : These are the parameters related to how the csv trace data is passed in, and the simulation safe-guards. These parameters are required and will be overriden with defaults if not provided with warnings. \n",
```
Comment (apavlen) [inline examples/using_vasim.ipynb:174]: added a bit of explanation how it works

File: src/vasim/recommender/cluster_state_provider/ConfigStateConstants.py [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+#  Copyright (c) Microsoft Corporation.
+# --------------------------------------------------------------------------
+#
+
+# General Config Defaults
+DEFAULT_WINDOW = 60   # minutes
+DEFAULT_LAG = 15
+DEFAULT_MAX_CPU_LIMIT = 20
+DEFAULT_MIN_CPU_LIMIT = 1
+RECOVERY_TIME = 15   # minutes
```
[the remark this answers is not in the document]
Comment (apavlen) [reply src/vasim/recommender/cluster_state_provider/ConfigStateConstants.py:14]: well it's a start. there is probably more magic left :)

**Comment by apavlen** [comment]
> One thing - in the notebook there is a section _"prediction_config : this relates to the window of...."_ where it describes the prediction config and each field. Can you please add a line about what the new `"enabled": true,`does? Thanks!

Done!

**Review by ksaur — APPROVED** [review]
LGTM!

**Pull request merged**